compiler-cli/src/dependencies.rs

```diff
@@ -492,38 +516,230 @@ fn resolve_versions<Telem: Telemetry>(
     telemetry: &Telem,
 ) -> Result<Manifest, Error> {
     telemetry.resolving_package_versions();
-    let resolved = hex::resolve_versions(
+    let dependencies = config.dependencies_for(mode)?;
+    let locked = config.locked(manifest)?;
+
+    let mut version_requirements = HashMap::new();
+    let mut provided_packages = HashMap::new();
+    let mut provider_info = HashMap::new();
+    for (name, recipe) in dependencies.into_iter() {
+        let version = match recipe {
+            Recipe::Hex { version } => version,
+            Recipe::Path { path } => {
+                provide_local_package(&name, &path, &mut provider_info, &mut provided_packages)?
+            }
+            Recipe::Git { git } => {
+                provide_git_package(&name, &git, &mut provider_info, &mut provided_packages)?
+            }
+        };
+        let _ = version_requirements.insert(name, version);
+    }
+
+    let resolved = dependency::resolve_versions(
         PackageFetcher::boxed(runtime.clone()),
-        mode,
-        config,
-        manifest,
+        provided_packages.clone(),
+        config.name.to_string(),
+        version_requirements.into_iter(),
+        &locked,
     )?;
-    let packages = runtime.block_on(future::try_join_all(
-        resolved
-            .into_iter()
-            .map(|(name, version)| lookup_package(name, version)),
-    ))?;
+
+    let provided_package_requirements = provided_packages
+        .into_iter()
+        .map(|(name, package)| {
+            (
+                name,
+                package.releases[0].requirements.keys().cloned().collect(),
+            )
+        })
+        .collect();
```

Comment:
Looks like we can avoid taking ownership of the data and doing all the cloning here as it's only used by reference.

Why is it that we have the structure `package.releases[0]` when we constructed this packages and so earlier had their requirements to hand? If we don't put them into a vector we can statically ensure they are present rather than doing a runtime check.

Reply:
This is because the `Package` provided by `hexpm` expects a vector of releases. We can't port the package struct over, but We could perhaps be working with releases directly here instead of packages.

Reply:
This is a map of `Package` rather than `Release` because we assembled it to pass into `resolve_versions`. I just to bite the bullet and make a `LocalPackage` struct which all these other things can be properly derived from.